File: CompetitiveProgramming/BitMagic/lonely_integer.rb
#!/usr/bin/env ruby

# We could solve this problem by using a HashMap to store the occurrences of
# each number and whichever key has '1' as its value is the lonely integer.
#
# We could also sort the array, and when we find a number different from both,
# the previous and the next one, then that's our lonely integer.
#
# Or we could use some bit magic. We are guaranteed that every number occurs
# twice sans one (the lonely integer). If we XOR any number with itself, then
# it becomes zero. So, if we XOR everything, all the repeated numbers will cancel
# themselves out, leaving us with the answer we seek: The Lonely Integer.

def lonely_integer(arr)
  loner = 0
  arr.each { |num| loner ^= num }
  loner
end

# Available Test Cases:
#
# Input: [1, 1, 2]
# Output: 2
#
# Input: [0, 0, 1, 2, 1]
# Output: 2

File: InterviewPrep/lonely_integer.c
#include <stdio.h>
#include "tools/competitiveinput.h"

int lonely_integer(int, int *);

int main(void)
{
    int n = parse_int(ltrim(rtrim(readline())));
    char **a_temp = split_string(rtrim(readline()));
    int *a = malloc(n * sizeof(int));

    for (int i = 0; i < n; i++)
    {
        int a_item = parse_int(*(a_temp + i));
        *(a + i) = a_item;
    }

    int result = lonely_integer(n, a);
    printf("%d\n", result);

    free(a_temp);
    free(a);
    return 0;
}

int lonely_integer(int arr_count, int *arr)
{
    // We can solve this one with some bit magic!
    //
    // If we xor any number with itself, then the result is 0. In other words,
    // if we xor the entire input, all the pairs will end up cancelling themselves,
    // leaving us with only the lonely integer!

    int lonely = 0;

    for (int i = 0; i < arr_count; i++)
    {
        lonely ^= *(arr + i);
    }

    return lonely;
}
